[async_tp] Support mm+rs with scatter_dim matmul K by sharding B

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #163069
* #163068
* __->__ #162794

Current state: Shape mismatch failure when mm+rs on the last mm scatter dim.

Adding separate path to handle lastdim for aten.mm, scaled_mm should be handled similarly, but needs additional PR.
So disabling scaled_mm case with filter matmul function.

Adding inductor.config for this change that is True by default for fast debuggability of new path.

cc @H-Huang @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @ezyang @msaroufim @dcci @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben

IvanKobzarev committed:
[async_tp] Support mm+rs with scatter_dim matmul K by sharding B

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support mm+rs with scatter_dim matmul K by sharding B"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support mm+rs with scatter_dim matmul K by sharding B"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support mm+rs with scatter_dim matmul K by sharding B"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[async_tp] Support mm+rs with scatter_dim matmul K by sharding B"




cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci

[ghstack-poisoned]

fegin reviewed (approved):
LGTM, but I'm wondering do we actually want a config or always enable this. Async TP ops already have logic to dispatch to a cublas version if kernels are small.

fegin commented:
The failing test is not related as async TP doesn't use NVSHMEM.

IvanKobzarev commented:
> LGTM, but I'm wondering do we actually want a config or always enable this. Async TP ops already have logic to dispatch to a cublas version if kernels are small.

Agree, I think I should remove config.

IvanKobzarev committed:
Update on "[async_tp] Support mm+rs with scatter_dim matmul K by sharding B"


Current state: Shape mismatch failure when mm+rs on the last mm scatter dim.

Adding separate path to handle lastdim for aten.mm, scaled_mm should be handled similarly, but needs additional PR.
So disabling scaled_mm case with filter matmul function.

Adding inductor.config for this change that is True by default for fast debuggability of new path.

cc H-Huang awgu wanchaol fegin fduwjj wz337 wconstab d4l3k pragupta ezyang msaroufim dcci voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov coconutruben

[ghstack-poisoned]
